src/basic/chattr-util.c

```diff
@@ -172,3 +173,56 @@ int read_attr_at(int dir_fd, const char *path, unsigned *ret) {
 
         return read_attr_fd(fd, ret);
 }
+
+int get_proj_id(int fd, uint32_t *ret) {
+        struct fsxattr attrs;
+
+        if (ioctl(fd, FS_IOC_FSGETXATTR, &attrs) < 0)
+                return -errno;
+
+        if (attrs.fsx_projid > 0)
+                *ret = attrs.fsx_projid;
+
+        return 0;
+}
+
+int set_proj_id(const char *path, uint32_t proj_id) {
+        struct fsxattr attrs;
+
+        int fd = xopenat(AT_FDCWD, path, O_RDONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW);
+        if (fd < 0)
+                return fd;
+
+        if (ioctl(fd, FS_IOC_FSGETXATTR, &attrs) < 0)
+                return -errno;
+
+        if (attrs.fsx_projid == proj_id)
+                return 0;
+
+        attrs.fsx_projid = proj_id;
+
+        return RET_NERRNO(ioctl(fd, FS_IOC_FSSETXATTR, &attrs));
+}
+
+static int set_proj_id_cb(
+                RecurseDirEvent event,
+                const char *path,
+                int dir_fd,
+                int inode_fd,
+                const struct dirent *de,
+                const struct statx *sx,
+                void *userdata) {
+        uint32_t proj_id = *(uint32_t *) userdata;
+        return set_proj_id(path, proj_id);
```

Comment:
please do not go by path here. You get an O_PATH inode fd already as input here in `input_fd`, use that. it might be necessary to convert it to a regular fd (i.e. non-O_PATH) first before you can execute FS_IOC_FSGETXATTR on it though.

use the RECURSE_DIR_INODE_FD flag

Reply:
and before applying the proj id, please ensure you only do this for regular files and for dirs. opening the other file types usually doesn't work (in case of sockets) or will connect you with a different fs, where the xattr ioctls are not going to work on anyway...